persistence/sql/process_local_queue.go

```diff
@@ -0,0 +1,178 @@
+	if err != nil {
+		return nil, err
+	}
+
+	waitingQueues, err := persistence.NewStateExecutionWaitingQueuesFromBytes(prcRow.StateExecutionWaitingQueues)
+	if err != nil {
+		return nil, err
+	}
+
+	// merge waitingQueues.UnconsumedMessages into request.Messages to consume
+	// waitingQueues.UnconsumedMessages need to be first to ensure the consumption order
+	messages := append(waitingQueues.UnconsumedMessages, request.Messages...)
+	waitingQueues.ClearUnconsumedMessages()
+
+	for _, message := range messages {
+		assignedStateExecutionIdString, hasFinishedWaiting := waitingQueues.Consume(message)
+
+		if assignedStateExecutionIdString == nil {
+			continue
+		}
+
+		consumedMessages, ok := assignedStateExecutionIdToMessagesMap[*assignedStateExecutionIdString]
+		if !ok {
+			consumedMessages = []persistence.LocalQueueMessageInfoJson{}
+		}
+		consumedMessages = append(consumedMessages, message)
+		assignedStateExecutionIdToMessagesMap[*assignedStateExecutionIdString] = consumedMessages
+
+		if hasFinishedWaiting {
+			finishedStateExecutionIdToPreviousVersionMap[*assignedStateExecutionIdString] = 0
+		}
+	}
+
+	prcRow.StateExecutionWaitingQueues, err = waitingQueues.ToBytes()
+	if err != nil {
+		return nil, err
+	}
+
+	err = tx.UpdateProcessExecution(ctx, *prcRow)
+	if err != nil {
+		return nil, err
+	}
+
+	// Step 2: update state execution rows for assignedStateExecutionIdToMessagesMap
+	for assignedStateExecutionIdString, messages := range assignedStateExecutionIdToMessagesMap {
+		stateExecutionId, err := persistence.NewStateExecutionIdFromString(assignedStateExecutionIdString)
+		if err != nil {
+			return nil, err
+		}
+
+		stateRow, err := tx.SelectAsyncStateExecutionForUpdate(ctx, extensions.AsyncStateExecutionSelectFilter{
+			ProcessExecutionId: prcRow.ProcessExecutionId,
+			StateId:            stateExecutionId.StateId,
+			StateIdSequence:    stateExecutionId.StateIdSequence,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		commandResults, err := persistence.BytesToCommandResults(stateRow.WaitUntilCommandResults)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, message := range messages {
+			dedupIdString := message.DedupId.String()
+			payload := message.Payload
+			commandResults.LocalQueueResults = append(commandResults.LocalQueueResults, xdbapi.LocalQueueMessage{
+				QueueName: message.QueueName,
+				DedupId:   &dedupIdString,
+				Payload:   &payload,
+			})
+		}
+
+		stateRow.WaitUntilCommandResults, err = persistence.FromCommandResultsToBytes(commandResults)
+		if err != nil {
+			return nil, err
+		}
+
+		err = tx.UpdateAsyncStateExecution(ctx, *stateRow)
+		if err != nil {
+			return nil, err
+		}
+
+		_, ok := finishedStateExecutionIdToPreviousVersionMap[assignedStateExecutionIdString]
+		if ok {
+			finishedStateExecutionIdToPreviousVersionMap[assignedStateExecutionIdString] = stateRow.PreviousVersion
+		}
+	}
+
+	// Step 3: handle finished wait_until task for finishedStateExecutionIdToPreviousVersionMap
```

Comment:
Currently, only the local queue commands are considered. If to support other commands, then we need to check the completion type and other command results in this place.